### DISCUSSION TITLE
Define manuSpecificTuya attributes in cluster.js genBasic? 0xfffe, 0xffde, 0xffe2, 0xffe4

### DISCUSSION THREAD

**Discussion by AlexKay88** [Koenkk/zigbee2mqtt#12305] [Devices]
During sniffing of several Tuya devices connected to Tuya gateway I observe that Tuya gateway does pretty much always the same initialization sequence when new devices join.
The following is a example `configure` section for a device to imitate Tuya initialization:

```
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint = device.getEndpoint(1);
            // Initialization sequence with Tuya gateway
            await endpoint.read('genBasic', ['manufacturerName', 'zclVersion', 'appVersion', 'modelId', 'powerSource', 'tuya0xfffe']);
            await endpoint.write('genBasic', {'tuya0xffde': 19}); 
            await endpoint.command('manuSpecificTuya', 'dataQuery', {});
            await endpoint.command('manuSpecificTuya', 'mcuVersionRequest', {'seq': 0x0002});
            try {
                await endpoint.read('genBasic', ['tuya0xffe2']);
            } catch (err) {/* do nothing */}
            try {
                await endpoint.read('genBasic', ['tuya0xffe4']);
            } catch (err) {/* do nothing */}
        },
```
`tuya0xfffe, tuya0xffde, tuya0xffe2, tuya0xffe4` are currently not defined in `cluster.js`.
`tuya0xffde` must be defined to be able to use write command `endpoint.write('genBasic', {'tuya0xffde': 19});`.
Others may be read wthout prior definition with e.g. `endpoint.read('genBasic', [ 0xFFFE ]);` 

Would it make sense to add those attributes to `cluster.js` for future use?
This could be used for a future manuSpecificTuya initialization function to make it easier to support new Tuya devices, which prerequisite parts of this initialization sequence for proper function.
Some devices already need special initialization as you can see here:
  [Enables reporting of physical state changes](https://github.com/Koenkk/zigbee2mqtt/issues/9057#issuecomment-1007742130)

`zigbee-herdsman/dist/zcl/definition/cluster.js`
```
const Cluster = {
    genBasic: {
        ID: 0,
        attributes: {
            tuya0xffde: { ID: 65502, type: dataType_1.default.uint8 },
            tuya0xffe2: { ID: 65506, type: dataType_1.default.uint8 }, 
            tuya0xffe4: { ID: 65508, type: dataType_1.default.uint8 }, 
            clusterRevision: { ID: 65533, type: dataType_1.default.uint16 }, // taken from Tuya developer doc
            tuya0xfffe: { ID: 65534, type: dataType_1.default.enum8 },
        },
```
Note: above `dataTypes `already verified with sniffer trace

**Comment by Koenkk** [comment]
I don't think defining it in cluster.js adds a lot of value. What does add a lot of value is defining such function in https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/lib/tuya.js . Also I don't think everything is needed to make the device fully work; as far as I saw just `await endpoint.read('genBasic', ['manufacturerName', 'zclVersion', 'appVersion', 'modelId', 'powerSource', 'tuya0xfffe']);` is enough

**Reply by AlexKay88** [reply]
Do you think about a function `tuyaDeviceInitialization` which does at least
`await endpoint.read('genBasic', ['manufacturerName', 'zclVersion', 'appVersion', 'modelId', 'powerSource', 'tuya0xfffe']);`
and optionally further `endpoint.read(...)` and `endpoint.write(...)` ?

I have to improve my skills, but with help of some other experienced code contributors I would give it a try...

**Reply by Koenkk** [reply]
> and optionally further endpoint.read(...) and endpoint.write(...)

Have you found any devices where this is necessary for?

**Reply by AlexKay88** [reply]
Not for the devices I have in use here.
Looking at other devices I find only 3 devices in `devices/tuya.js` which have additional configure steps:
```
            await endpoint.write('genOnOff', {'tuyaOperationMode': 1});
            await endpoint.read('genOnOff', ['tuyaOperationMode']);
            try {
                await endpoint.read(0xE001, [0xD011]);
            } catch (err) {/* do nothing */}
```
Can't judge if `endpoint.write('genOnOff', {'tuyaOperationMode': 1});` and `endpoint.read(0xE001, [0xD011]);` is really needed for proper operation.

At the moment there are only 12 devices which require
            `endpoint.read('genBasic', ['manufacturerName', 'zclVersion', 'appVersion', 'modelId', 'powerSource', 0xfffe]);`

From my opinion too few devices which require `endpoint.read` and `endpoint.write`.
For the other 12 devices which require `endpoint.read('genBasic', ['manufacturerName', 'zclVersion', 'appVersion', 'modelId', 'powerSource', 'tuya0xfffe']);` this is only one additional code line, thus a function would not add so much value.

**Reply by Koenkk** [reply]
Agree, lets keep it as is.
